In [2]:
# Import Pandas.
import pandas as pd
import numpy as np
# Import the movies data set
nc1 = pd.read_excel('national_categories.xlsx')

# Read CSV file from the current working directory.
ad1 = pd.read_csv('actual_duration.csv')

# Read CSV file from the current working directory.
ar1 = pd.read_csv('appointments_regional.csv')



# Data imported correctly?
print(nc1.columns)
print(nc1.dtypes)
print(nc1.shape)
print(ad1.columns)
print(ad1.dtypes)
print(ad1.shape)
print(ar1.columns)
print(ar1.dtypes)
print(ar1.shape)


Index(['appointment_date', 'icb_ons_code', 'sub_icb_location_name',
       'service_setting', 'context_type', 'national_category',
       'count_of_appointments', 'appointment_month'],
      dtype='object')
appointment_date         datetime64[ns]
icb_ons_code                     object
sub_icb_location_name            object
service_setting                  object
context_type                     object
national_category                object
count_of_appointments             int64
appointment_month                object
dtype: object
(817394, 8)
Index(['sub_icb_location_code', 'sub_icb_location_ons_code',
       'sub_icb_location_name', 'icb_ons_code', 'region_ons_code',
       'appointment_date', 'actual_duration', 'count_of_appointments'],
      dtype='object')
sub_icb_location_code        object
sub_icb_location_ons_code    object
sub_icb_location_name        object
icb_ons_code                 object
region_ons_code              object
appointment_date             object
actual_du

In [3]:
# Values of appointment date before scaling.
ad1['appointment_date']



0         01-Dec-21
1         01-Dec-21
2         01-Dec-21
3         01-Dec-21
4         01-Dec-21
            ...    
137788    30-Jun-22
137789    30-Jun-22
137790    30-Jun-22
137791    30-Jun-22
137792    30-Jun-22
Name: appointment_date, Length: 137793, dtype: object

#  Between what dates were appointments scheduled?

In [4]:
# Determining the minimum and maximum dates in the ad DataFrame.

# max = Returns max of a column
print(ad1['appointment_date'].max())

# min = Returns min of a column
print(ad1['appointment_date'].min())




31-May-22
01-Apr-22


# grouping the dates with max and min count of appointments on the day in the actual duration dataframe.


In [5]:
# Determining the minimum and maximum dates in the ad DataFrame.
# grouping the dates with max and min count of appointments on the day.



ad1.groupby('appointment_date').count_of_appointments.agg(
    dates_max=('max'),
    dates_min=('min'),
)

dates_max  dates_min
appointment_date                      
01-Apr-22             11470         62
01-Dec-21             12568         69
01-Feb-22             13255         85
01-Jan-22               228          1
01-Jun-22             12015         64
...                     ...        ...
30-May-22             13542         73
31-Dec-21              8251         45
31-Jan-22             14664         96
31-Mar-22             11282         74
31-May-22             13004         95

[212 rows x 2 columns]

In [6]:
# Determining the minimum and maximum dates in the nc DataFrame.

# max = Returns max of a column
print(nc1['appointment_date'].max())

# min = Returns min of a column
print(nc1['appointment_date'].min())



2022-06-30 00:00:00
2021-08-01 00:00:00


# grouping the dates with max and min count of appointments on the day in the national categories(nc1) dataframe

In [7]:
# Determining the minimum and maximum dates in the nc DataFrame.
# grouping the dates with max and min count of appointments on the day.



nc1.groupby('appointment_date').count_of_appointments.agg(
    dates_max=('max'),
    dates_min=('min'),
)

dates_max  dates_min
appointment_date                      
2021-08-01              607          1
2021-08-02            13150          1
2021-08-03            12657          1
2021-08-04            12344          1
2021-08-05            10882          1
...                     ...        ...
2022-06-26              355          1
2022-06-27            16242          1
2022-06-28            15424          1
2022-06-29            14107          1
2022-06-30            13932          1

[334 rows x 2 columns]

# Which service setting was the most popular for NHS North West London from 1 January to 1 June 2022?

In [8]:
# Concatenate the three DataFrames.
nc2_ad2_ar2 = pd.concat([nc1, ad1,ar1], axis=0)

# View the new DataFrame.
print(nc2_ad2_ar2.shape)
nc2_ad2_ar2.fillna(0)

(1552008, 16)


appointment_date icb_ons_code  \
0       2021-08-02 00:00:00    E54000050   
1       2021-08-02 00:00:00    E54000050   
2       2021-08-02 00:00:00    E54000050   
3       2021-08-02 00:00:00    E54000050   
4       2021-08-02 00:00:00    E54000050   
...                     ...          ...   
596816                    0    E54000050   
596817                    0    E54000050   
596818                    0    E54000050   
596819                    0    E54000050   
596820                    0    E54000050   

                             sub_icb_location_name       service_setting  \
0       NHS North East and North Cumbria ICB - 00L  Primary Care Network   
1       NHS North East and North Cumbria ICB - 00L                 Other   
2       NHS North East and North Cumbria ICB - 00L      General Practice   
3       NHS North East and North Cumbria ICB - 00L      General Practice   
4       NHS North East and North Cumbria ICB - 00L      General Practice   
...                                            ...                   ...   
596816                                           0                     0   
596817                                           0                     0   
596818                                           0                     0   
596819                                           0                     0   
596820                                           0                     0   

                  context_type                       national_category  \
0       Care Related Encounter  Patient contact during Care Home Round   
1       Care Related Encounter                         Planned Clinics   
2       Care Related Encounter                              Home Visit   
3       Care Related Encounter              General Consultation Acute   
4       Care Related Encounter            Structured Medication Review   
...                        ...                                     ...   
596816                       0                                       0   
596817                       0                                       0   
596818                       0                                       0   
596819                       0                                       0   
596820                       0                                       0   

        count_of_appointments appointment_month sub_icb_location_code  \
0                           3           2021-08                     0   
1                           7           2021-08                     0   
2                          79           2021-08                     0   
3                         725           2021-08                     0   
4                           2           2021-08                     0   
...                       ...               ...                   ...   
596816                      1           2022-06                     0   
596817                      1           2022-06                     0   
596818                      1           2022-06                     0   
596819                      1           2022-06                     0   
596820                      1           2022-06                     0   

       sub_icb_location_ons_code region_ons_code actual_duration  \
0                              0               0               0   
1                              0               0               0   
2                              0               0               0   
3                              0               0               0   
4                              0               0               0   
...                          ...             ...             ...   
596816                         0               0               0   
596817                         0               0               0   
596818                         0               0               0   
596819                         0               0               0   
596820                         0               0       

In [9]:
# Convert the data type of column 'Date' from string (YYYY/MM/DD) to datetime64
nc2_ad2_ar2["appointment_month"] =  pd.to_datetime(nc2_ad2_ar2["appointment_month"], format="%Y/%m/%d")
nc2_ad2_ar2.dtypes


appointment_date                             object
icb_ons_code                                 object
sub_icb_location_name                        object
service_setting                              object
context_type                                 object
national_category                            object
count_of_appointments                         int64
appointment_month                    datetime64[ns]
sub_icb_location_code                        object
sub_icb_location_ons_code                    object
region_ons_code                              object
actual_duration                              object
appointment_status                           object
hcp_type                                     object
appointment_mode                             object
time_between_book_and_appointment            object
dtype: object

In [10]:
# Create group 1.
group1 = nc2_ad2_ar2.groupby(['service_setting', 'count_of_appointments','appointment_month'])

# View the DataFrame.
group1
group1.sum()

/var/folders/vf/ymq2hz2x7lnb9nnlr6_ycvfc0000gn/T/ipykernel_32983/1940968481.py:6: FutureWarning: Dropping invalid columns in DataFrameGroupBy.add is deprecated. In a future version, a TypeError will be raised. Before calling .add, select only columns which should be valid for the function.
  group1.sum()


icb_ons_code  \
service_setting           count_of_appointments appointment_month                                                      
Extended Access Provision 1                     2021-08-01         E54000050E54000050E54000050E54000050E54000050E...   
                                                2021-09-01         E54000050E54000050E54000050E54000050E54000050E...   
                                                2021-10-01         E54000050E54000050E54000050E54000050E54000050E...   
                                                2021-11-01         E54000050E54000050E54000050E54000050E54000050E...   
                                                2021-12-01         E54000050E54000050E54000050E54000050E54000050E...   
...                                                                                                              ...   
Unmapped                  4290                  2021-09-01                                                 E54000027   
                          4298                  2021-08-01                                                 E54000027   
                          4329                  2021-10-01                                                 E54000027   
                          4384                  2021-10-01                                                 E54000027   
                          4387                  2021-10-01                                                 E54000027   

                                                                                               sub_icb_location_name  \
service_setting           count_of_appointments appointment_month                                                      
Extended Access Provision 1                     2021-08-01         NHS North East and North Cumbria ICB - 00LNHS ...   
                                                2021-09-01         NHS North East and North Cumbria ICB - 00LNHS ...   
                                                2021-10-01         NHS North East and North Cumbria ICB - 00LNHS ...   
                                                2021-11-01         NHS North East and North Cumbria ICB - 00LNHS ...   
                                                2021-12-01         NHS North East and North Cumbria ICB - 00LNHS ...   
...                                                                                                              ...   
Unmapped                  4290                  2021-09-01                         NHS North West London ICB - W2U3Z   
                          4298                  2021-08-01                         NHS North West London ICB - W2U3Z   
                          4329                  2021-10-01                         NHS North West London ICB - W2U3Z   
                          4384                  2021-10-01                         NHS North West London ICB - W2U3Z   
                          4387                  2021-10-01                         NHS North West London ICB - W2U3Z   

                                                                                                        context_type  \
service_setting           count_of_appointments appointment_month                                                      
Extended Access Provision 1                     2021-08-01         Inconsistent MappingInconsistent MappingCare R...   
                                                2021-09-01         Inconsistent MappingCare Related EncounterInco...   
                                                2021-10-01         Inconsistent MappingCare Related EncounterInco...   
                                                2021-11-01         Inconsistent MappingCare Related EncounterInco...   
                                                2021-12-01         Inconsistent MappingInconsistent MappingCare R...   
...                                                                                                              ...   
Unmapped              

# Which month had the highest number of appointments?

# the month of october 2021-10-01 had the highest number of count of appointments with 4387 unmapped services in the NHS North West London area.

# What was the total number of records per month?

In [27]:
# Create group 1.
group1 = nc2_ad2_ar2.groupby(['service_setting', 'count_of_appointments','appointment_month',]) \
.agg(['sum', 'mean']) \
.sort_values('appointment_month',ascending=False)
group1.fillna(0)

sub_icb_location_code  \
                                                                                    sum   
service_setting           count_of_appointments appointment_month                         
General Practice          1602                  2022-06-01                            0   
                          742                   2022-06-01                            0   
                          740                   2022-06-01                            0   
Primary Care Network      647                   2022-06-01                            0   
General Practice          5048                  2022-06-01                            0   
...                                                                                 ...   
Primary Care Network      200                   2021-08-01                            0   
General Practice          1237                  2021-08-01                            0   
Unmapped                  514                   2021-08-01                            0   
General Practice          3896                  2021-08-01                            0   
Extended Access Provision 1                     2021-08-01                            0   

                                                                        \
                                                                  mean   
service_setting           count_of_appointments appointment_month        
General Practice          1602                  2022-06-01         0.0   
                          742                   2022-06-01         0.0   
                          740                   2022-06-01         0.0   
Primary Care Network      647                   2022-06-01         0.0   
General Practice          5048                  2022-06-01         0.0   
...                                                                ...   
Primary Care Network      200                   2021-08-01         0.0   
General Practice          1237                  2021-08-01         0.0   
Unmapped                  514                   2021-08-01         0.0   
General Practice          3896                  2021-08-01         0.0   
Extended Access Provision 1                     2021-08-01         0.0   

                                                                  sub_icb_location_ons_code  \
                                                                                        sum   
service_setting           count_of_appointments appointment_month                             
General Practice          1602                  2022-06-01                                0   
                          742                   2022-06-01                                0   
                          740                   2022-06-01                                0   
Primary Care Network      647                   2022-06-01                                0   
General Practice          5048                  2022-06-01                                0   
...                                                                                     ...   
Primary Care Network      200                   2021-08-01                                0   
General Practice          1237                  2021-08-01                                0   
Unmapped                  514                   2021-08-01                                0   
General Practice          3896                  2021-08-01                                0   
Extended Access Provision 1                     2021-08-01                                0   

                                                                        \
                                                                  mean   
service_setting           count_of_appointments appointment_month        
General Practice          1602                  2022-06-01         0.0   
                          742                   2022-06-01         0.0   
                          740                   2022-

# the total record per month for general practice recorded in june 2022 was 5048
# the total record per month for primary care network in june 2022 was 647

